commit 9a396274520148c3c7b6dc916d751035f690b308
Author: Peter Michael Green <plugwash@debian.org>
Date:   Thu Jan 10 13:50:19 2019 +0000

    Avoid installing shared libraries to man-page directories.

diff --git a/Makefile b/Makefile
index b73505a..9e43eab 100644
--- a/Makefile
+++ b/Makefile
@@ -106,9 +106,9 @@ install:	$(ALL)
 	if which python2; then python2 setup.py install $(PYINSTALLARGS); fi
 	if which python3; then python3 setup.py install $(PYINSTALLARGS); fi
 	install -m 0755 -d                             $(DESTDIR)$(mandir)/man1
-	install -m 0644 *.1                            $(DESTDIR)$(mandir)/man1
+	install -m 0644 p*.1                           $(DESTDIR)$(mandir)/man1
 	install -m 0755 -d                             $(DESTDIR)$(mandir)/man3
-	install -m 0644 *.3                            $(DESTDIR)$(mandir)/man3
+	install -m 0644 p*.3                           $(DESTDIR)$(mandir)/man3
 ifeq ($(DESTDIR),)
 	ldconfig
 endif
