--- a/Makefile
+++ b/Makefile
@@ -31,7 +31,7 @@
 
 LL3      = -L. -lpigpiod_if2 -pthread -lrt
 
-prefix = /usr/local
+prefix = /usr
 exec_prefix = $(prefix)
 bindir = $(exec_prefix)/bin
 includedir = $(prefix)/include
@@ -83,7 +83,7 @@ clean:
 ifeq ($(DESTDIR),)
   PYINSTALLARGS =
 else
-  PYINSTALLARGS = --root=$(DESTDIR)
+  PYINSTALLARGS = --install-layout=deb --root=$(DESTDIR)
 endif
 
 install:	$(ALL)
